source/Scope.cpp

```diff
@@ -4,18 +4,42 @@
 namespace slang {
 
 static const Scope emptyScope;
-const Scope* const Scope::Empty = &emptyScope;
+const Scope *const Scope::Empty = &emptyScope;
 
-void Scope::add(const Symbol* symbol) {
-    ASSERT(symbol);
-    table.emplace(symbol->name, symbol);
+Scope::Scope() : parentScope{Scope::Empty} {}
+
+bool Scope::add(const Symbol *symbol) {
+        ASSERT(symbol);
+    list.emplace_back(symbol);
+    return table.emplace(symbol->name, symbol).second;
+}
+
+const Symbol *Scope::lookup(StringRef name, bool local) const {
+    for (auto scope = this; scope && scope != Scope::Empty; scope = scope->parent()) {
```

Comment:
I think we shouldn't require all scopes to point to Scope::Empty... it should be fine for the parent pointer to just be null (which is the default when a scope is first constructed), in which case we don't need any special knowledge of Scope::Empty.